bdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,766 @@
+            Ok(())
+        }
+    }
+
+    /// Populate an update [`SparseChain`] with transactions (and associated block positions) that
+    /// contain and spend the given `outpoints`.
+    fn populate_with_outpoints(
+        &self,
+        update: &mut SparseChain<P>,
+        outpoints: impl Iterator<Item = OutPoint>,
+    ) -> Result<HashMap<Txid, Transaction>, InternalError> {
+        let tip = update
+            .latest_checkpoint()
+            .expect("update must atleast have one checkpoint");
+        let mut full_txs = HashMap::new();
+        for outpoint in outpoints {
+            for (tx_height, tx) in self.find_spending_tx(outpoint)? {
+                let txid = tx.txid();
+                full_txs.insert(txid, tx);
+                if tx_height.is_confirmed() && tx_height > TxHeight::Confirmed(tip.height) {
+                    continue;
+                }
+                let tx_pos = self.get_chain_position(txid, tx_height)?;
+                if let Err(failure) = update.insert_tx(txid, tx_pos) {
+                    match failure {
+                        sparse_chain::InsertTxError::TxTooHigh { .. } => {
+                            unreachable!(
+                                "we should never encounter this as we ensured height <= tip"
+                            );
+                        }
+                        sparse_chain::InsertTxError::TxMovedUnexpectedly { .. } => {
+                            return Err(InternalError::Reorg);
+                        }
+                    }
+                }
+            }
+        }
+        Ok(full_txs)
+    }
+
+    /// Populate an update [`SparseChain`] with transactions (and associated block positions) from
+    /// the given `txids`.
+    fn populate_with_txids(
+        &self,
+        update: &mut SparseChain<P>,
+        txids: impl Iterator<Item = Txid>,
+    ) -> Result<(), InternalError> {
+        let tip = update
+            .latest_checkpoint()
+            .expect("update must have atleast one checkpoint");
+        for txid in txids {
+            let tx_height = match self.get_txid_status(txid)? {
+                Some(height) => height,
+                None => continue,
+            };
+            if tx_height.is_confirmed() && tx_height > TxHeight::Confirmed(tip.height) {
+                continue;
+            }
+            let tx_pos = self.get_chain_position(txid, tx_height)?;
+            if let Err(failure) = update.insert_tx(txid, tx_pos) {
+                match failure {
+                    sparse_chain::InsertTxError::TxTooHigh { .. } => {
+                        unreachable!("we should never encounter this as we ensured height <= tip");
+                    }
+                    sparse_chain::InsertTxError::TxMovedUnexpectedly { .. } => {
+                        return Err(InternalError::Reorg);
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn populate_with_txs<'a>(
```

Comment:
many methods here are only called in one place. Please inline them to their call location so the code is easier to reason about.

Reply:
Done in bbff340